MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c

```diff
@@ -48,6 +97,210 @@ StatusToHexString (
+  // they are not garunteed to be in any particular order
+  //
+  CertList = (EFI_SIGNATURE_LIST *)Db;
+
+  //
+  // Iterate through the DB entries
+  // The DB is a list of EFI_SIGNATURE_LISTs
+  //
+  while ((DbSize > 0) && (DbSize >= CertList->SignatureListSize)) {
+    //
+    // If the signature type is not X509, skip it
+    //
+    if (CompareGuid (&CertList->SignatureType, &gEfiCertX509Guid)) {
+      //
+      // Otherwise, we have a certificate list and now we need to iterate through the certificates
+      //
+      CertData  = (EFI_SIGNATURE_DATA *)((UINT8 *)CertList + sizeof (EFI_SIGNATURE_LIST) + CertList->SignatureHeaderSize);
+      CertCount = (CertList->SignatureListSize - sizeof (EFI_SIGNATURE_LIST) - CertList->SignatureHeaderSize) / CertList->SignatureSize;
+
+      //
+      // Iterate over each certificate in the list
+      //
+      for (Index = 0; Index < CertCount; Index++) {
+        //
+        // grab the certificate and its size
+        //
+        Cert     = CertData->SignatureData;
+        CertSize = CertList->SignatureSize - sizeof (EFI_GUID);
+
+        //
+        // First lets ensure that the size of the certificate is at least as large as the certificate we are looking for
+        // Each offset in the certficate entry plus the size of the string must be less than the size of the certificate
+        //
+        if (((Certificate->OrganizationOffset + sizeof (Certificate->Organization)) <= CertSize) &&
+            ((Certificate->CommonNameOffset + sizeof (Certificate->CommonName)) <= CertSize) &&
+            ((Certificate->SignatureOffset + sizeof (Certificate->Signature)) <= CertSize))
+        {
+          //
+          // If the certificate is the one we are looking for, return EFI_SUCCESS
+          //
+          if ((CompareMem (Cert + Certificate->OrganizationOffset, Certificate->Organization, Certificate->OrganizationLength) == 0) &&
+              (CompareMem (Cert + Certificate->CommonNameOffset, Certificate->CommonName, Certificate->CommonNameLength) == 0) &&
+              (CompareMem (Cert + Certificate->SignatureOffset, Certificate->Signature, sizeof (Certificate->Signature)) == 0))
+          {
+            Status = EFI_SUCCESS;
+            Found  = TRUE;
+            goto Exit;
+          }
+        }
+      }
+    }
+
+    //
+    // Regardless of the signature type, we need to move to the next signature list
+    // Subtract the size of the current signature list from the total size of the DB
+    // Then move the pointer to the next signature list
+    //
+    DbSize  -= CertList->SignatureListSize;
+    CertList = (EFI_SIGNATURE_LIST *)((UINT8 *)CertList + CertList->SignatureListSize);
+  }
+
+  //
+  // If we get here, we did not find the certificate in the DB, however, we did not encounter any errors
+  //
+  Status = EFI_SUCCESS;
+
+Exit:
+
+  if (OptionalStatus != NULL) {
+    *OptionalStatus = Status;
+  }

```

Comment:
If I understand the flow correctly, `Db` is no longer needed and the buffer can be freed (on success or through an error path).

```suggestion
Exit:

  if (Db != NULL) {
    mBS->FreePool (Db);
  }

  if (OptionalStatus != NULL) {
    *OptionalStatus = Status;
  }

```

Reply:
This is correct, somehow this got dropped when I was rebasing commits trying to clean up the commits. Thanks!